src/app/modules/bounce/bounce.controller.ts

```diff
@@ -23,11 +24,28 @@ export const handleSns: RequestHandler<
   // so we never fail on malformed input. The response code is meaningless since
   // it is meant to go back to AWS.
   try {
-    const isValid = await snsService.isValidSnsRequest(req.body)
-    if (!isValid) {
-      return res.sendStatus(StatusCodes.FORBIDDEN)
+    const isValid = await BounceService.isValidSnsRequest(req.body)
+    if (!isValid) return res.sendStatus(StatusCodes.FORBIDDEN)
+
+    const notification: IEmailNotification = JSON.parse(req.body.Message)
+    BounceService.logEmailNotification(notification)
+    if (
+      BounceService.extractEmailType(notification) !== EmailType.AdminResponse
+    ) {
+      return res.sendStatus(StatusCodes.OK)
+    }
+    const bounceDoc = await BounceService.getUpdatedBounceDoc(notification)
+    // Missing headers in notification
+    if (!bounceDoc) return res.sendStatus(StatusCodes.OK)
+    if (bounceDoc.areAllPermanentBounces()) {
+      await BounceService.deactivateFormFromBounce(bounceDoc)
+    }
+    if (bounceDoc.isCriticalBounce()) {
+      const emailRecipients = await BounceService.notifyAdminOfBounce(bounceDoc)
```

Comment:
extract the check out of `notifyAdminOfBounce` since `notifyAdminOfBounce` is performing an action, and we should not be conflating a query (hasNotified) with a command (notifyAdminOfBounce).

```typescript
    if (bounceDoc.isCriticalBounce() && !bounceDoc.hasNotified()) {
      const emailRecipients = await BounceService.notifyAdminOfBounce(bounceDoc)
```